lib/util/http/github.ts

```diff
@@ -201,68 +206,67 @@ export class GithubHttp extends Http<GithubHttpOptions, GithubHttpOptions> {
+          let lastPage = +linkHeader.last.page;
+          // istanbul ignore else: needs a test
+          if (!process.env.RENOVATE_PAGINATE_ALL && opts.paginate !== 'all') {
+            lastPage = Math.min(pageLimit, lastPage);
+          }
+          const pageNumbers = Array.from(
+            new Array(lastPage),
+            (x, i) => i + 1
+          ).slice(1);
+          const queue = pageNumbers.map(
+            (page) => (): Promise<HttpResponse<T>> => {
+              const nextUrl = new URL(linkHeader.next.url, baseUrl);
+              nextUrl.search = '';
+              nextUrl.searchParams.set('page', page.toString());
+              return this.request<T>(
+                nextUrl,
+                { ...opts, paginate: false },
+                okToRetry
+              );
             }
-            const pageNumbers = Array.from(
-              new Array(lastPage),
-              (x, i) => i + 1
-            ).slice(1);
-            const queue = pageNumbers.map(
-              (page) => (): Promise<HttpResponse> => {
-                const nextUrl = new URL(linkHeader.next.url, baseUrl);
-                delete nextUrl.search;
-                nextUrl.searchParams.set('page', page.toString());
-                return this.request(
-                  nextUrl,
-                  { ...opts, paginate: false },
-                  okToRetry
-                );
+          );
+          const pages = await pAll(queue, { concurrency: 5 });
+          if (opts.paginationField && is.plainObject(result.body)) {
+            const paginatedResult = result.body[opts.paginationField];
+            if (is.array<T>(paginatedResult)) {
+              for (const nextPage of pages) {
+                if (is.plainObject(nextPage.body)) {
+                  const nextPageResults = nextPage.body[opts.paginationField];
+                  if (is.array<T>(nextPageResults)) {
+                    paginatedResult.push(...nextPageResults);
+                  }
+                }
+              }
+            }
+          } else if (is.array<T>(result.body)) {
+            for (const nextPage of pages) {
+              if (is.array<T>(nextPage.body)) {
+                result.body.push(...nextPage.body);
               }
-            );
-            const pages = await pAll(queue, { concurrency: 5 });
-            if (opts.paginationField) {
-              result.body[opts.paginationField] = result.body[
-                opts.paginationField
-              ].concat(
-                ...pages
-                  .filter(Boolean)
-                  .map((page) => page.body[opts.paginationField])
-              );
-            } else {
-              result.body = result.body.concat(
-                ...pages.filter(Boolean).map((page) => page.body)
-              );
             }
           }
         }
       }
+      return result;
     } catch (err) {
-      handleGotError(err, url, opts);
+      throw handleGotError(err, url, opts);
```

Comment:
Why do we need this?

Reply:
In order to make `request` return only `Response` or throw (and never return `null`), we have to explicitly (re)throw something from the catch block.

Reply:
Also, `handleGotError()` that returns `Error` explicitly rather than `never` hypothetically should help to avoid silent exits from the function.